File: 2022/august/07-08-2022/count-vowels-permutation.py
from functools import lru_cache

class Solution:
    def countVowelPermutation(self, n: int) -> int:

        # A mapper mapped characters to their next characters
        mapper = {
            "": ["a", "e", "i", "o", "u"],
            "a": "e",
            "e": ["a", "i"],
            "i": ["a", "e", "o", "u"],
            "o": ["i", "u"],
            "u": ["a"],
        }

        @lru_cache(None)
        def dp(n, c):

            # If n == 1, we have reach base case and thus, return 1
            if n == 1:
                return 1

            # Initialize the total to 0
            total = 0

            # Recursively solve sub-problems until n is reduced to 1
            for char in mapper[c]:
                total = (total + dp(n - 1, char)) % 1000000007

            return total

        # Add 1 to n since we started with empty string instead of recursively called dp on each vowel
        return dp(n + 1, "")

File: 2022/august/07-08-2022/count-vowels-permutation.rs
impl Solution {
    pub fn count_vowel_permutation(n: i32) -> i32 {
        const C: i64 = 10_i64.pow(9) + 7;
        let mut old = [1_i64; 5];
        let mut new = [1_i64; 5];
        for _ in 1..n {
            new[0] = (old[1] + old[2] + old[4]) % C; //a
            new[1] = (old[0] + old[2]) % C; //e
            new[2] = (old[1] + old[3]) % C; //i
            new[3] = old[2] % C; //o
            new[4] = (old[2] + old[3]) % C; //u
            old = new;
        }
        ((old[0] + old[1] + old[2] + old[3] + old[4]) % C) as i32
    }
}
